### PR TITLE
feat(store-engine,store-engine-bro-fs,store-engine-dexie,store-engine-fs,store-engine-sqleet): Add generic for primary key

### PR DESCRIPTION
**Pull request by ffflorian**
## Pull Request Checklist

- [x] My code is covered by tests
- [x] I will [merge the PR as breaking change](https://github.com/wireapp/wire-web-packages/wiki/Releases#create-major-release), if the API contract changes

### CODE REVIEW

File: packages/store-engine-dexie/src/index.test.ts [hunk 1]
```diff
@@ -51,8 +51,8 @@ describe('IndexedDBEngine', () => {
   });
 
   afterEach(done => {
-    if (engine && engine.db) {
-      engine.db.close();
+    if (engine && engine['db']) {
```
Comment (bennyn) [inline packages/store-engine-dexie/src/index.test.ts]: Can't we do `engine.db` anymore?

File: packages/store-engine-dexie/src/index.test.ts [hunk 2]
```diff
@@ -61,7 +61,7 @@ describe('IndexedDBEngine', () => {
   describe('init', () => {
     it('resolves with the database instance to which the records will be saved.', async () => {
       engine = new IndexedDBEngine();
-      const instance = await engine.init(STORE_NAME);
+      const instance: Dexie = await engine.init(STORE_NAME);
```
Comment (bennyn) [inline packages/store-engine-dexie/src/index.test.ts]: Do we need to cast to `Dexie` or can we work with the generic?

File: packages/store-engine-dexie/src/index.test.ts [hunk 3]
```diff
@@ -141,6 +141,31 @@ describe('IndexedDBEngine', () => {
...
+      });
+      await db.open();
+
+      const entity = {
+        testValue: 'value',
+      };
+
+      type ExpectedResult = typeof entity & {primaryKey: number};
+
+      const primaryKey = await engine.updateOrCreate<number>(TABLE_NAME, undefined as any, entity);
```
Comment (bennyn) [inline packages/store-engine-dexie/src/index.test.ts]:
```suggestion
      const primaryKey = await engine.updateOrCreate<number | undefined>(TABLE_NAME, undefined, entity);
```

Can we do this to get rid off `any`?

File: packages/store-engine-dexie/src/index.ts [hunk 4]
```diff
@@ -145,19 +157,19 @@ export class IndexedDBEngine implements CRUDEngine {
...
 
-  public async readAllPrimaryKeys(tableName: string): Promise<string[]> {
+  public async readAllPrimaryKeys<PrimaryKey = string>(tableName: string): Promise<PrimaryKey[]> {
     const keys = await this.db
-      .table(tableName)
+      .table<PrimaryKey>(tableName)
       .toCollection()
       .keys();
-    return keys.map(key => `${key}`);
+    return keys.map(key => (key as any) as PrimaryKey);
```
Comment (bennyn) [inline packages/store-engine-dexie/src/index.ts]: Is it possible to type the `key`?

File: packages/store-engine-fs/src/index.ts [hunk 5]
```diff
@@ -82,14 +82,18 @@ export class FileEngine implements CRUDEngine {
...
+    tableName: string,
+    primaryKey: PrimaryKey,
+    entity: EntityType,
+  ): Promise<PrimaryKey> {
     if (entity) {
       const filePath = this.resolvePath(tableName, primaryKey);
       if (typeof entity === 'object') {
         try {
-          entity = JSON.stringify(entity);
+          entity = JSON.stringify(entity) as any;
```
Comment (bennyn) [inline packages/store-engine-fs/src/index.ts]: `JSON.stringify(entity)` is a string, so we might need to do `EntityType = Object | string` if possible?

File: packages/store-engine-sqleet/src/index.ts [hunk 6]
```diff
@@ -167,12 +167,16 @@ export class SQLeetEngine implements CRUDEngine {
...
+    tableName: string,
+    primaryKey: PrimaryKey,
+    entity: EntityType,
+  ): Promise<PrimaryKey> {
     if (!entity) {
       const message = `Record "${primaryKey}" cannot be saved in "${tableName}" because it's "undefined" or "null".`;
       throw new RecordTypeError(message);
     }
-    const {columns, values} = this.buildValues(tableName, entity);
+    const {columns, values} = this.buildValues(tableName, entity as any);
```
Comment (bennyn) [inline packages/store-engine-sqleet/src/index.ts]: Any way to kill `any` or to narrow it down to `Object` or `EntityType`?

File: packages/store-engine-dexie/src/index.test.ts (hunk 1 above)
Comment (ffflorian) [reply packages/store-engine-dexie/src/index.test.ts]: No, `db` is private.

File: packages/store-engine-dexie/src/index.test.ts (hunk 2 above)
Comment (ffflorian) [reply packages/store-engine-dexie/src/index.test.ts]: `init()` has no generic type and it returns `any`. So to have types in the tests, I added this cast.

**Pull request merged**